 <h1 align='center'> The most attractive app types for users</h1>



> ### Introduction

 - **What is the project about?** 
 
    The project aims to analysis apps data to check what kind of free app is more attractive for users download and install in their devices.
 
 ---
 
 - **What is the project goal?**
 
     Identifying what kind of app is more attractive for users we can focus on those apps to increase in-apps ads profit/revenue.

> ### Project origin

This project is a Guided Project: Profitable App Profiles for the App Store and Google Pay Markets from <a href="https://www.dataquest.io/">Dataquest</a>
    
Despite being a guided project, it's a nice Data Science project to work with two different datasets, data cleaning, analysis and business intelligence.

> ### Datasets

* Google Play Apps - August, 2018([Download can be done here](https://dq-content.s3.amazonaws.com/350/googleplaystore.csv))

* App Store Apps - July, 2017 ([Download can be done here](https://dq-content.s3.amazonaws.com/350/AppleStore.csv))

In [ ]:
### Defining a explore_data function to print rows and give 
### how many rows and cols the dataset has
def explore_data(dataset, start, end, rows_and_columns=False):
    dataset_slice = dataset[start:end]    
    for row in dataset_slice:
        print(row)
        print('\n') # adds a new (empty) line after each row

    if rows_and_columns:
        print('Number of rows:', len(dataset))
        print('Number of columns:', len(dataset[0]))

In [ ]:
### Opening csv files
opened_android = open('googleplaystore.csv', encoding='utf8')
opened_ios = open('AppleStore.csv', encoding='utf8')

### Transforming to iterable object and then casting to 
### list of lists
from csv import reader
android = list(reader(opened_android))
ios = list(reader(opened_ios))

### Exploring data
explore_data(android, 1, 3, True)
print("\n")
explore_data(ios, 1, 3, True)



['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', 'January 15, 2018', '2.0.0', '4.0.3 and up']


Number of rows: 10842
Number of columns: 13


['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


Number of rows: 7198
Number of columns: 16


### Play Store Apps dataset info (Android)
---
<a id="table-android"></a>

| Column Name | Description |
| :----:      |  :----:     |
| App | Application name|
| Category | Category the app belongs to|
| Rating | Overall user rating of the app|
| Reviews | Number of user reviews for the app|
| Size | Size of the app |
| Installs | Number of user downloads/installs for the app|
| Type | Paid or free|
| Price | Price of the app|
| Content rating| Age group the app is targeted at - Children/Mature 21+/Adult|
| Genres | An app can belong to multiple genres (apart from its main category) |
| Last Updated | Date when the app was last updated on Play Store |
| Current Ver | Current version of the app on Play Store |
| Android Ver | Min required Android version |


* **Number of rows: 10842**
* **Number of cols: 13**

### App Store Apps dataset info (iOS)
---
<a id="table-ios"></a>

| Column name | Description |
| :----:      | :----:      |
| "id"| App ID|
| "track_name"|App Name|
| "size_bytes"|Size (in Bytes)|
| "currency"|Currency Type|
| "price"|Price amount|
| "ratingcounttot"| User Rating counts (for all version)|
| "ratingcountver"| User Rating counts (for current version)|
| "user_rating"| Average User Rating value (for all version)|
| "userratingver"| Average User Rating value (for current version)|
| "ver"| Latest version code|
| "cont_rating"| Content Rating|
| "prime_genre"| Primary Genre |
| "sup_devices.num"| Number of supporting devices |
| "ipadSc_urls.num"| Number of screenshots showed for display|
| "lang.num"| Number of supported languages|
| "vpp_lic"| Vpp Device Based Licensing Enabled|
| "id" | App ID|

* **Number of rows: 7198**
* **Number of cols: 16**

* In Play Store apps dataset, row index 10473 has missing value and misplaced columns as we can see below:

In [ ]:
print(android[10471])
print("\n")
print(android[10472])
print("\n")
print("->", android[10473])
print("\n")
print(android[10474])

['Jazz Wi-Fi', 'COMMUNICATION', '3.4', '49', '4.0M', '10,000+', 'Free', '0', 'Everyone', 'Communication', 'February 10, 2017', '0.1', '2.3 and up']


['Xposed Wi-Fi-Pwd', 'PERSONALIZATION', '3.5', '1042', '404k', '100,000+', 'Free', '0', 'Everyone', 'Personalization', 'August 5, 2014', '3.0.0', '4.0.3 and up']


-> ['Life Made WI-Fi Touchscreen Photo Frame', '1.9', '19', '3.0M', '1,000+', 'Free', '0', 'Everyone', '', 'February 11, 2018', '1.0.19', '4.0 and up']


['osmino Wi-Fi: free WiFi', 'TOOLS', '4.2', '134203', '4.1M', '10,000,000+', 'Free', '0', 'Everyone', 'Tools', 'August 7, 2018', '6.06.14', '4.4 and up']


* To fix this issue we need to remove the data linked to this index

In [ ]:
""" 
Removing index 10473

First, let's print to check the values at index 10473, 
and then, after removing it, let's print again. 
"""
print("Old value at [10473] ->", android[10473],'\n')
del android[10473]
print("New value at [10473] ->", android[10473])

Old value at [10473] -> ['Life Made WI-Fi Touchscreen Photo Frame', '1.9', '19', '3.0M', '1,000+', 'Free', '0', 'Everyone', '', 'February 11, 2018', '1.0.19', '4.0 and up'] 

New value at [10473] -> ['osmino Wi-Fi: free WiFi', 'TOOLS', '4.2', '134203', '4.1M', '10,000,000+', 'Free', '0', 'Everyone', 'Tools', 'August 7, 2018', '6.06.14', '4.4 and up']


### Creating a function to check whether the dataset has duplicated rows or not

* The function will receive a dataset and a indexCol as argument and will return a list of app's names that are duplicated. If the list gets a `len()` == 0, it means that doesn't have duplicated rows


* It checks for duplicated rows based on indexCol passed as argument. For our purpose, we'll be using app's name column (Android - 'App' - index 0; iOS - 'track_name' - index 1)

In [ ]:
def checkDuplicatedRows(dataset, indexCol):
    duplicatedApps = []
    uniqueApps = []
    
    for row in dataset[1:]:        
        app_name = row[indexCol]
        
        if(app_name in uniqueApps):
            duplicatedApps.append(app_name)
        else:
            uniqueApps.append(app_name)
    
    return [duplicatedApps, uniqueApps]

In [ ]:
android_apps = checkDuplicatedRows(android, 0)
print("Duplicated rows size:",len(android_apps[0]))
print("\n")
print("Some examples of repeated app's entry:", android_apps[0][:10])

Duplicated rows size: 1181


Some examples of repeated app's entry: ['Quick PDF Scanner + OCR FREE', 'Box', 'Google My Business', 'ZOOM Cloud Meetings', 'join.me - Simple Meetings', 'Box', 'Zenefits', 'Google Ads', 'Google My Business', 'Slack']


In [ ]:
ios_apps = checkDuplicatedRows(ios, 1)
print("Duplicated rows size:",len(ios_apps[0]))
print("\n")
print("Some examples of repeated app's entry:", ios_apps[0])

Duplicated rows size: 2


Some examples of repeated app's entry: ['Mannequin Challenge', 'VR Roller Coaster']


* ### Criterion for duplicated rows (iOS dataset)

Below we can see the repeated rows/apps

In [ ]:
repeated_names = ['Mannequin Challenge', 'VR Roller Coaster']
print("iOS apps header:", ios[0], "\n")


for app in ios:
    name = app[1]    
    if(name in repeated_names):
        print(app, "\n")

iOS apps header: ['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic'] 

['1173990889', 'Mannequin Challenge', '109705216', 'USD', '0.0', '668', '87', '3.0', '3.0', '1.4', '9+', 'Games', '37', '4', '1', '1'] 

['952877179', 'VR Roller Coaster', '169523200', 'USD', '0.0', '107', '102', '3.5', '3.5', '2.0.0', '4+', 'Games', '37', '5', '1', '1'] 

['1178454060', 'Mannequin Challenge', '59572224', 'USD', '0.0', '105', '58', '4.0', '4.5', '1.0.1', '4+', 'Games', '38', '5', '1', '1'] 

['1089824278', 'VR Roller Coaster', '240964608', 'USD', '0.0', '67', '44', '3.5', '4.0', '0.81', '4+', 'Games', '38', '0', '1', '1'] 



For this case we can use 'ver' column (Version) to check which app is the most recent and then we'll be able to create a cleaned dataset. Steps to accomplish this:

* Find the 'ver' column index.


* Loop through dataset and compare those apps to see which one have the latest version.


* **(!)** Here we used some trick with regex to get some part of version string. Some version string is formatted as 'x.x.x' so these '.' throws a error when casting to float. Using regex we were able to get only the first two digits and cast it to float type to do some comparision.


* After identifying which app has the latest version, we save it to a dict to append to ios_clean list using `extend()` function. `extend()` allow us to add multiple values to a list.

In [ ]:
ver_index = ios[0].index('ver')
ios_clean = []
correct_rows = {}
latest_version = {}


import re

for app in ios[1:]:
    name = app[1]
    
    if(name in repeated_names):
        
        '''
        Little hack here:
            Using regex to get part of string version that can
            be casted to float
        '''
        result = re.search('\d.\d', app[ver_index])
        print(result.group())
        
        version = float(result.group())
        
        # If name already exists but value version is < 
        # Set new version value to key and update var version value
        if(name in latest_version and 
           latest_version[name] < version):            
            latest_version[name] = version
            correct_version[name] = app
            version = 0
            
        # If name already exists but value version is >
        # Keep key-value and update var version value
        elif(name in latest_version and
            latest_version[name] > version):
            version = 0
            
        # Else (it does not exist yet)
        # just add to dict
        else:
            latest_version[name] = version
            correct_rows[name] = app
    else:
        ios_clean.append(app)
            
print(latest_version)
print("\n")
print(correct_rows)
print("\n")

ios_clean.extend(list(correct_rows.values()))
print(ios_clean[-1])
print(ios_clean[-2])
print(ios_clean[-3])
            

1.4
2.0
1.0
0.8
{'Mannequin Challenge': 1.4, 'VR Roller Coaster': 2.0}


{'Mannequin Challenge': ['1173990889', 'Mannequin Challenge', '109705216', 'USD', '0.0', '668', '87', '3.0', '3.0', '1.4', '9+', 'Games', '37', '4', '1', '1'], 'VR Roller Coaster': ['952877179', 'VR Roller Coaster', '169523200', 'USD', '0.0', '107', '102', '3.5', '3.5', '2.0.0', '4+', 'Games', '37', '5', '1', '1']}


['952877179', 'VR Roller Coaster', '169523200', 'USD', '0.0', '107', '102', '3.5', '3.5', '2.0.0', '4+', 'Games', '37', '5', '1', '1']
['1173990889', 'Mannequin Challenge', '109705216', 'USD', '0.0', '668', '87', '3.0', '3.0', '1.4', '9+', 'Games', '37', '4', '1', '1']
['977965019', 'みんなのお弁当 by クックパッド ~お弁当をレシピ付きで記録・共有~', '51174400', 'USD', '0.0', '0', '0', '0.0', '0.0', '1.4.0', '4+', 'Food & Drink', '37', '0', '1', '1']


* ### Criterion for duplicated rows (Android dataset)

It's not a good practice just removing duplicated rows. We need to apply some criterions to determine which rows/registries keep for our analysis. Therefore, looking at datasets' info tables we could assume some signifcant column to guide us removing duplicated rows.

For Android analysis, we'll be using 'Review' column to remove the duplicated rows. The higher the number of reviews, the more recent the data should be.  

In [ ]:
print("Android dataset 'Review' column index:", 
      android[0].index('Reviews'))

reviews_max = {}

for app in android[1:]:
    
    name = app[0] # getting app's name
    n_reviews = float(app[3]) # getting app's review
    
    if(name in reviews_max and reviews_max[name] < n_reviews):
        reviews_max[name] = n_reviews
        
    elif(name not in reviews_max):
        reviews_max[name] = n_reviews

print("Dictionary size after Review column cleaning:", len(reviews_max))  
print("\n")
print("Size android dataset:",
      len(android[1:]))
print("Size android duplicated rows:",
      len(android_apps[0]))
correct_size = float(len(android[1:])-len(android_apps[0]))
print("Correct size after cleaning will be:", correct_size)

Android dataset 'Review' column index: 3
Dictionary size after Review column cleaning: 9659


Size android dataset: 10840
Size android duplicated rows: 1181
Correct size after cleaning will be: 9659.0


In [ ]:
android_clean = []
already_added = []

for app in android[1:]:
    name = app[0]
    n_reviews = float(app[3])
    """
    It's necessary to add app's name to already_added list if it is
    not in there yet because we can get some duplicated rows that 
    havethe same Reviews column value and it would make us add a 
    value twice in our android_clean list
    """
    if(n_reviews == reviews_max[name] and 
      name not in already_added):
        
        android_clean.append(app)
        already_added.append(name)

### After cleaning, android_clean list should have 9,659 rows
print("android_clean size:", len(android_clean), "rows")

android_clean size: 9659 rows


 ### Writing a function that checks if a app name is from english alphabet or not 


* Based on ASCII table, any char value over 127 doesn't belong to English alphabet


* Our function should be able to take a string as parameter and looping through it check if any char is > 127


* The function should be able to tolerate some emojis in app's name. Here we can consider until 3 times. 

In [ ]:
def check_english_alpha(string):
    count_tolerate = 0
    for ch in string:
        if(ord(ch) > 127): 
            count_tolerate += 1
            if(count_tolerate == 3):
                return False
    return True

### Time to validate the function with some data test
test = ['Instagram', 
        '爱奇艺PPS -《欢乐颂2》电视剧热播', 
        'Docs To Go™ Free Office Suite',
        'Instachat 😜']

for data in test:
    print(check_english_alpha(data))
    

True
False
True
True


After testing the function that will indicate which apps are in english and see that it's working, let's use on datasets!

In [ ]:
### iOS 
print('iOS app name index:', ios[0].index('track_name'))
ios_clean_english = []
for app in ios_clean:
    name = app[1]
    english_name = check_english_alpha(name)
    if(english_name):
        ios_clean_english.append(app)
        
print("Cleaned iOS dataset all apps names:", len(ios_clean))
print("Cleaned + English iOS dataset:", len(ios_clean_english))
print('\n')
print(ios_clean_english[0:2])
print(ios_clean[0:2])

iOS app name index: 1
Cleaned iOS dataset all apps names: 7195
Cleaned + English iOS dataset: 6153
[['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1'], ['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']]
[['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1'], ['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']]


In [ ]:
### Android
print('android app name index:', android[0].index('App'))
android_clean_english = []
for app in android_clean:
    name = app[1]
    english_name = check_english_alpha(name)
    if(english_name):
        android_clean_english.append(app)
        
print("Cleaned android dataset all apps names:", len(android_clean))
print("Cleaned + English android dataset:", len(android_clean_english))
print('\n')
print(android_clean[0:2])
print(android_clean_english[0:2])

android app name index: 0
Cleaned android dataset all apps names: 9659
Cleaned + English android dataset: 9659
[['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up'], ['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']]
[['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up'], ['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']]


### Isolating only free apps for analysis

In [ ]:
### iOS

ios_free_apps = []
price_index = ios[0].index('price')

for app in ios_clean_english:
    price = app[price_index]
    # Making sure that any variation of free/0/0.0 is acceptable
    if(price == '0' or price == '0.0'or price == 'free'):
        ios_free_apps.append(app)

explore_data(ios_free_apps, 0, 6, True)

['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'Games', '38', '5', '18', '1']


['420009108', 'Temple Run', '65921024', 'USD', '0.0', '1724546', '3842', '4.5', '4.0', '1.6.2', '9+', 'Games', '40', '5', '1', '1']


['284035177', 'Pandora - Music & Radio', '130242560', 'USD', '0.0', '1126879', '3594', '4.0', '4.5', '8.4.1', '12+', 'Music', '37', '4', '1', '1']


['429047995', 'Pinterest', '74778624', 'USD', '0.0', '1061624', '1814', '4.5', '4.0', '6.26', '12+', 'Social Networking', '37', '5', '27', '1']


Number of rows: 3201
Number of columns: 16


In [ ]:
### Android

android_free_apps = []
# Type columns shows if an app is Free or paid
price_index = android[0].index('Type')

for app in android_clean_english:
    price = app[price_index]
    # Making sure that any variation of free is acceptable (ignore.case == True)
    if( re.search('free', price, re.IGNORECASE) ):
        android_free_apps.append(app)
        

explore_data(android_free_apps, 0, 10, True)

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


['Pixel Draw - Number Art Coloring Book', 'ART_AND_DESIGN', '4.3', '967', '2.8M', '100,000+', 'Free', '0', 'Everyone', 'Art & Design;Creativity', 'June 20, 2018', '1.1', '4.4 and up']


['Paper flowers instructions', 'ART_AND_DESIGN', '4.4', '167', '5.6M', '50,000+', 'Free', '0', 'Everyone', 'Art & Design', 'March 26, 2017', '1.0', '2.3 and up']


['Smoke Effect Photo Maker - Smoke Editor', 'ART_AND_DESIGN', '3.8', '178', '19M', '50,000+

* **(!)** Another regex trick used above for android dataset. In `if()` condition inside for loop, we used regex with `re.IGNORECASE` flag that allow us to search for 'Free' value independently if it is correctly written or mistyped like 'free' or 'fRee' and so on 

> ### Validation strategy for an app idea

* The idea behind this is to find out the app profile that fits both App Store and Google Play.

* To accomplish this we need to create a frequency table, for both datasets cleaned, and identify which genre is more popular among users.

* We need inspect columns and see which one is best choice for this purpose. We can also check datasets info tables (<a href="#table-android">Android</a> or <a href="#table-ios">iOS</a>) to see what represents each column.

In [ ]:
explore_data(ios, 0, 2, False)
explore_data(android, 0, 2, False)

['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']




* Looking at what is printed above and datasets info tables, for iOS we can use **'prime_genre'** to check which genre is the most common among ios users and for the most common among android users we can use **'Category'**. 


* **(!)** We'll be using 'Category' at android dataset instead of 'Genre' because 'Genre' is not necessary showing the real category/genre of an app, and as info table says 'Category' is the main category/classification of an app.

In [ ]:
'''
    freq_table function is a function that takes in two parameters
    dataset - list of lists and index - integer. The function returns
    the frequency table (as dictionary) for any column.
'''
def freq_table(dataset, index, header = False):
    f_table = {}
    
    # Making sure that no header is included    
    if(header):
        dataset = dataset[1:]
    for app in dataset:
        key = app[index]
        # If exists, +1
        if(key in f_table):
            f_table[key] += 1
        # Not exists, create and count 1    
        else:
            f_table[key] = 1
            
    # Transforming into percentages
    for key in f_table:
        f_table[key] = round(float((f_table[key]/len(dataset))*100), 2)
        
    return f_table

'''
    display_table takes in any dataset - list of lists and any 
    index - integer and display, in a sorted way, the frequency
    table created by freq_table function
'''
def display_table(dataset, index):
    table = freq_table(dataset, index)
    table_display = []
    for key in table:
        key_val_as_tuple = (table[key], key)
        table_display.append(key_val_as_tuple)

    table_sorted = sorted(table_display, reverse = True)
    for entry in table_sorted:
        print(entry[1], ':', entry[0])

In [ ]:
print("iOS 'prime_genre' index", ios[0].index('prime_genre'))
print("Android 'Category' index", android[0].index('Category'))

iOS 'prime_genre' index 11
Android 'Category' index 1


In [ ]:
# iOS visualization of frequency table
print('iOS Frequency Table', '\n')
display_table(ios_free_apps, 11)

iOS Frequency Table 

Games : 58.23
Entertainment : 7.84
Photo & Video : 5.0
Education : 3.69
Social Networking : 3.31
Shopping : 2.59
Utilities : 2.47
Sports : 2.16
Music : 2.06
Health & Fitness : 2.03
Productivity : 1.75
Lifestyle : 1.56
News : 1.34
Travel : 1.25
Finance : 1.09
Weather : 0.87
Food & Drink : 0.81
Reference : 0.53
Business : 0.53
Book : 0.37
Navigation : 0.19
Medical : 0.19
Catalogs : 0.12


> ### iOS Frequency Table Analysis

* The frequency table above shows that **Games** has more than a half of presence in iOS apps (~58.23%). It means that **Games** is the most common app for iOS users but it does not mean that those apps has the greatest number of users among iOS apps. 

* We can see that **Games** is followed by **Entertainment** category. It has around 7.84% of presence among iOS apps. 

* So far we can say that for iOS apps, the most common apps that fits around 66,07% (2/3 of apps) are apps related to helping users spending their time. And this argument is reinforced by the top 2 apps category in frequency table above. Although the data is not current, nowadays we have as example for **Entertainment** category apps that are streaming platform like Netflix, Amazon Prime, Disney+ and so on. For the top-tier category, **Games**, we have famous examples like HearthStone, Pokemon Go and Clash of Clans. 

In [ ]:
# Android visualization of frequency table
print('Android Frequency Table', '\n')
display_table(android_free_apps, 1)

Android Frequency Table 

FAMILY : 18.97
GAME : 9.7
TOOLS : 8.43
BUSINESS : 4.58
LIFESTYLE : 3.93
PRODUCTIVITY : 3.89
FINANCE : 3.68
MEDICAL : 3.52
SPORTS : 3.38
PERSONALIZATION : 3.31
COMMUNICATION : 3.23
HEALTH_AND_FITNESS : 3.07
PHOTOGRAPHY : 2.94
NEWS_AND_MAGAZINES : 2.83
SOCIAL : 2.65
TRAVEL_AND_LOCAL : 2.32
SHOPPING : 2.25
BOOKS_AND_REFERENCE : 2.18
DATING : 1.85
VIDEO_PLAYERS : 1.8
MAPS_AND_NAVIGATION : 1.42
FOOD_AND_DRINK : 1.24
EDUCATION : 1.17
ENTERTAINMENT : 0.95
LIBRARIES_AND_DEMO : 0.93
AUTO_AND_VEHICLES : 0.92
HOUSE_AND_HOME : 0.82
WEATHER : 0.8
EVENTS : 0.71
PARENTING : 0.65
ART_AND_DESIGN : 0.65
COMICS : 0.63
BEAUTY : 0.6


* **(!)** We could have done this frequency table for android apps using also 'Genres' column but this columns references for app's secondary category. By curiosity means, let's see what 'Genres' column tell us about which genre/category is the most common among android apps.

In [ ]:
print('Android \'Genres\' column', 
      android[0].index('Genres'), 
      '\n')

print('Android Frequency Table', '\n')
display_table(android_free_apps, 9)

Android 'Genres' column 9 

Android Frequency Table 

Tools : 8.42
Entertainment : 6.09
Education : 5.39
Business : 4.58
Lifestyle : 3.92
Productivity : 3.89
Finance : 3.68
Medical : 3.52
Sports : 3.45
Personalization : 3.31
Communication : 3.23
Action : 3.09
Health & Fitness : 3.07
Photography : 2.94
News & Magazines : 2.83
Social : 2.65
Travel & Local : 2.31
Shopping : 2.25
Books & Reference : 2.18
Simulation : 2.07
Dating : 1.85
Arcade : 1.84
Video Players & Editors : 1.77
Casual : 1.75
Maps & Navigation : 1.42
Food & Drink : 1.24
Puzzle : 1.12
Racing : 0.99
Role Playing : 0.93
Libraries & Demo : 0.93
Auto & Vehicles : 0.92
Strategy : 0.91
House & Home : 0.82
Weather : 0.8
Events : 0.71
Adventure : 0.69
Comics : 0.62
Art & Design : 0.61
Beauty : 0.6
Parenting : 0.49
Card : 0.45
Trivia : 0.43
Casino : 0.43
Educational;Education : 0.39
Board : 0.38
Educational : 0.37
Education;Education : 0.35
Word : 0.26
Casual;Pretend Play : 0.24
Music : 0.2
Racing;Action & Adventure : 0.17
Puzzle;B

> ### Android Frequency Table Analysis

* For android dataset, we've got a balanced distribution, even looking at secondary category that is 'Genre' column.

* Looking at main category, 'Category' column, we have at first position apps that are related to **'FAMILY'**. These apps related with **'FAMILY'** category are apps that can be classified as **'FAMILY-FRIEND'** and we can see it in Google documentation [FAMILY-FRIEND](https://support.google.com/googleplay/answer/6209531?hl=en). We can have several apps in this category of **FAMILY-FRIEND**, so we can't define a single pattern of apps for this category, however they should be apps that is ok for elders, middle-aged, adults, youth and children to use.

* The next positions in the frequency table can give more insights about what kind of app is more common. Besides apps 'Games' at second position, that are for entertainment and have a good time, from 3rd postion until 12th position, we've got apps for self-development/personal use. We can see categories like: finance, productivity, comunication, health and these kind of stuff.

* Even the 'Genre' column, that have secondary category value for apps, shows that from 3rd position until 9th position, apps for self-development/personal use are most common in Google Play, it shows values like: Education, Business, Lifestyle, Productivity, Finance, Medical and Sports.

* So, looking all that android data and after this analysis, we could say that apps for entertainment and/or apps for self-improvement (Lifestyle, Productivity, Finance, Business, Education, Communication) are the most common apps in Google Play. 


> ### Most Common Apps x Greatest Users Apps

* Even though we can tell which app is more common/has more frequency in App Store or Google Play, it does not shows us the effectiveness about hitting many number of users. 

* We can find the number of users analysing the **'Installs'** column for android dataset but for ios we have to do some workaround and use the column **'rating_count_tot'**. (We need do this workaround because number of installations is missing for ios dataset).

In [ ]:
### iOS
'''
    The main idea here is for each key (app's genre) in freq_table
    we should account for each app in ios-free-english apps
    that is the same genre the total ratings (total) and 
    increment by 1 length of genre (len_genre). 
    After iterating all apps, we calculate the average 
    rating dividing total ratings by length of genre. 
'''
ios_freq_table = freq_table(ios_free_apps,  11)

for key_genre in ios_freq_table:
    total = 0
    len_genre = 0
    
    for app in ios_free_apps:
        genre_app = app[11]
        
        if(genre_app == key_genre):
            total += float(app[7])
            len_genre += 1
            
    avg = total/len_genre        
    print('App genre:', key_genre)
    print('Average users ratings:', avg)
    print('\n')

App genre: Social Networking
Average users ratings: 3.5943396226415096


App genre: Photo & Video
Average users ratings: 3.903125


App genre: Games
Average users ratings: 4.052038626609442


App genre: Music
Average users ratings: 3.946969696969697


App genre: Reference
Average users ratings: 3.8823529411764706


App genre: Health & Fitness
Average users ratings: 3.769230769230769


App genre: Weather
Average users ratings: 3.482142857142857


App genre: Utilities
Average users ratings: 3.569620253164557


App genre: Travel
Average users ratings: 3.4875


App genre: Shopping
Average users ratings: 3.9759036144578315


App genre: News
Average users ratings: 3.244186046511628


App genre: Navigation
Average users ratings: 3.8333333333333335


App genre: Lifestyle
Average users ratings: 3.48


App genre: Entertainment
Average users ratings: 3.547808764940239


App genre: Food & Drink
Average users ratings: 3.6346153846153846


App genre: Sports
Average users ratings: 3.0652173913043477


> # Some Conclusions - iOS

Catalogs apps have the highest average rating but they are the least frequent apps. Games have the second highest average rating and they are the most frequent apps, so we could assume that for App Store, iOS users, focusing on apps games would be the best option for in-apps adds. 

In [ ]:
#print(android[0])
android_freq_table = freq_table(android_free_apps, 1)

'''
    Before diving into code to account how many install an app's
    genre has, we should be aware of Intalls column values. 
    It has values acting like symbolic intervals, so we can see 
    values like 1,000,000+; 100,000+, 500+... We can't specifically 
    determine which value belongs to each genre, so we should 
    consider these symbolic values and and treat them to 
    remove special characters(+).
'''
for key_catg in android_freq_table:
    total_inst = 0
    len_genre = 0
    
    for app in android_free_apps:
        if(key_catg == app[1]):
            install = app[5].replace("+", "")
            install = install.replace(",", '')
            
            
            total_inst += float(install)
            len_genre += 1
            
    avg = total_inst/len_genre
    
    # Rounding 2 decimal points
    avg = round(avg, 2)
    
    
    print('App genre:', key_catg)
    print('Average users installs:', avg)
    print('\n')

App genre: ART_AND_DESIGN
Average users installs: 1952105.17


App genre: AUTO_AND_VEHICLES
Average users installs: 647317.82


App genre: BEAUTY
Average users installs: 513151.89


App genre: BOOKS_AND_REFERENCE
Average users installs: 8587351.86


App genre: BUSINESS
Average users installs: 1708215.91


App genre: COMICS
Average users installs: 803234.82


App genre: COMMUNICATION
Average users installs: 38322625.7


App genre: DATING
Average users installs: 854028.83


App genre: EDUCATION
Average users installs: 1825480.77


App genre: ENTERTAINMENT
Average users installs: 11640705.88


App genre: EVENTS
Average users installs: 253542.22


App genre: FINANCE
Average users installs: 1387692.48


App genre: FOOD_AND_DRINK
Average users installs: 1924897.74


App genre: HEALTH_AND_FITNESS
Average users installs: 4188821.99


App genre: HOUSE_AND_HOME
Average users installs: 1331540.56


App genre: LIBRARIES_AND_DEMO
Average users installs: 638503.73


App genre: LIFESTYLE
Average user

> # Some conclusions - Android

* Analysing data by number of installations show us a new and different perspective of apps that could reach more number of users. We've got 9 category apps that has 10,000,000+ installs:

1) App genre: COMMUNICATION
   Average users installs: 38,322,625.7

2) App genre: VIDEO_PLAYERS
   Average users installs: 24,573,948.25

3) App genre: SOCIAL
   Average users installs: 23,253,652.13

4) App genre: PHOTOGRAPHY
   Average users installs: 17,772,018.76

5) App genre: PRODUCTIVITY
   Average users installs: 16,738,957.55

6) App genre: GAME
   Average users installs: 15,551,995.89

7) App genre: TRAVEL_AND_LOCAL
   Average users installs: 13,984,077.71

8) App genre: ENTERTAINMENT
   Average users installs: 11,640,705.88

9) App genre: TOOLS
   Average users installs: 10,787,009.95
   
  The sum of these nine categories represents something around 172.624.986 installs. As a meaning of comparisons, this number overpass the Germany and UK population combined ([Countries population](https://pt.wikipedia.org/wiki/Lista_de_pa%C3%ADses_por_popula%C3%A7%C3%A3o)).
  
  So if focusing on android apps, our efforts for in-apps adds could be directed for apps linked to Entertainment (VIDEO-PLAYERS, PHOTOGRAPHY, GAMES, ENTERTAINMENT) or for apps linked to self-improvement/development (COMMUNICATION, PRODUCTIVITY). As an audacious step, aiming users from Entertainment and self-improvement/development, the efforts could be directed for social apps, and this category is also the third with the most number of users. Social apps help users to get entertained and also help them to improve their communication and sociability.